### ISSUE TITLE
[Bug]: matplotlib dependency not resolved for Python 3.8

### ISSUE THREAD

**Issue by bitnom** [microsoft/autogen#1167]
### Describe the bug

autogen's current pin of matplotlib appears to be erroneous.

While working on [PEP 621 & pdm](https://github.com/microsoft/autogen/pull/1155)

🔒 Lock failed
WARNING: Unable to find a resolution because the following dependencies don't work on all Python versions in the range of the project's `requires-python`: >=3.8,<3.13.
  python>=3.9 (from <Candidate matplotlib@3.8.1 from https://pypi.org/simple/matplotlib/>)
  python>=3.9 (from <Candidate matplotlib@3.8.2 from https://pypi.org/simple/matplotlib/>)

### Steps to reproduce

set min python version to 3.8 in pyproject.toml and run `pdm lock`.

### Expected Behavior

pdm.lock

### Screenshots and logs

```
➜ pdm lock
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping ipython@8.19.0 because it requires Python>=3.10 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of ipython that supports Python>=3.8,<3.13 will be used.
If you want to install ipython@8.19.0, narrow down the `requires-python` range to include this version. For example, ">=3.10,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping ipython@8.18.1 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of ipython that supports Python>=3.8,<3.13 will be used.
If you want to install ipython@8.18.1, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping ipython@8.18.0 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of ipython that supports Python>=3.8,<3.13 will be used.
If you want to install ipython@8.18.0, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping ipython@8.17.2 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of ipython that supports Python>=3.8,<3.13 will be used.
If you want to install ipython@8.17.2, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping ipython@8.17.1 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of ipython that supports Python>=3.8,<3.13 will be used.
If you want to install ipython@8.17.1, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping ipython@8.16.1 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of ipython that supports Python>=3.8,<3.13 will be used.
If you want to install ipython@8.16.1, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping ipython@8.15.0 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of ipython that supports Python>=3.8,<3.13 will be used.
If you want to install ipython@8.15.0, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping ipython@8.14.0 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of ipython that supports Python>=3.8,<3.13 will be used.
If you want to install ipython@8.14.0, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping ipython@8.13.2 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of ipython that supports Python>=3.8,<3.13 will be used.
If you want to install ipython@8.13.2, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping ipython@8.13.1 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of ipython that supports Python>=3.8,<3.13 will be used.
If you want to install ipython@8.13.1, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping matplotlib@3.8.2 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another
version of matplotlib that supports Python>=3.8,<3.13 will be used.
If you want to install matplotlib@3.8.2, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping matplotlib@3.8.1 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another
version of matplotlib that supports Python>=3.8,<3.13 will be used.
If you want to install matplotlib@3.8.1, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping networkx@3.2.1 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of networkx that supports Python>=3.8,<3.13 will be used.
If you want to install networkx@3.2.1, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping pre-commit@3.6.0 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another
version of pre-commit that supports Python>=3.8,<3.13 will be used.
If you want to install pre-commit@3.6.0, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping contourpy@1.2.0 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of contourpy that supports Python>=3.8,<3.13 will be used.
If you want to install contourpy@1.2.0, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping numpy@1.26.3 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of numpy that supports Python>=3.8,<3.13 will be used.
If you want to install numpy@1.26.3, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping numpy@1.26.2 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of numpy that supports Python>=3.8,<3.13 will be used.
If you want to install numpy@1.26.2, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping numpy@1.26.1 because it requires Python<3.13,>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, 
another version of numpy that supports Python>=3.8,<3.13 will be used.
If you want to install numpy@1.26.1, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping numpy@1.26.0 because it requires Python<3.13,>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, 
another version of numpy that supports Python>=3.8,<3.13 will be used.
If you want to install numpy@1.26.0, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping numpy@1.25.2 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of numpy that supports Python>=3.8,<3.13 will be used.
If you want to install numpy@1.25.2, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping numpy@1.25.1 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of numpy that supports Python>=3.8,<3.13 will be used.
If you want to install numpy@1.25.1, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
/home/user/.local/lib/python3.11/site-packages/pdm/resolver/providers.py:158: PackageWarning: Skipping numpy@1.25.0 because it requires Python>=3.9 but the project claims to work with Python>=3.8,<3.13. Instead, another 
version of numpy that supports Python>=3.8,<3.13 will be used.
If you want to install numpy@1.25.0, narrow down the `requires-python` range to include this version. For example, ">=3.9,<3.13" should work.
  return self.repository.find_candidates(
🔒 Lock failed
WARNING: Unable to find a resolution because the following dependencies don't work on all Python versions in the range of the project's `requires-python`: >=3.8,<3.13.
  python>=3.9 (from <Candidate matplotlib@3.8.1 from https://pypi.org/simple/matplotlib/>)
  python>=3.9 (from <Candidate matplotlib@3.8.2 from https://pypi.org/simple/matplotlib/>)
A possible solution is to change the value of `requires-python` in pyproject.toml to >=3.9,<3.13.
```

### Additional Information

autogen's current pin of matplotlib appears to be erroneous. To resolve this, I bumped the minimum Python version requirement to `>=3.9` in the PR. I imagine it may have also been possible to specify an older matplotlib. IMO 3.8 is at least one version too old, though that point is worthy of RFC.

**Comment by rickyloynd-microsoft** [comment]
@sonichi fyi

**Comment by bitnom** [comment]
Related discussion: [Drop support for Python 3.8 #1173](https://github.com/microsoft/autogen/discussions/1173)

**Comment by davorrunje** [comment]
The matplot lib is only used in notebooks and should be removed from dependencies of the project. Notebooks are examples of how to use autogen and as their number grows, their dependencies will eventually conflict. We should remove them from autogen dependencies and keep instructions on how to install them in notebooks. The main library should have as few dependencies as possible to be able to cover wide range of applications.

**Comment by bitnom** [comment]
Resolved in f3f6fc0
